File: sdk/clojure/src/raptoreum.clj
(ns raptoreum
  (:require [clj-http.client :as client]
            [cheshire.core :as json])
  (:import [java.util Base64]))

(defn- base64-encode [s]
  (.encodeToString (Base64/getEncoder) (.getBytes s "UTF-8")))

(defn make-client [& {:keys [host port user password use-ssl]
                      :or {host "127.0.0.1" port 8766 use-ssl false}}]
  (let [scheme (if use-ssl "https" "http")
        url (str scheme "://" host ":" port "/")
        auth (if (and user password (not= user "") (not= password ""))
               (str "Basic " (base64-encode (str user ":" password)))
               nil)]
    {:url url :auth auth}))

(defn request [client method & [params]]
  (let [payload (json/generate-string
                 {:jsonrpc "1.0"
                  :id "rtm-sdk-clojure"
                  :method method
                  :params (or params [])})
        headers (merge {"Content-Type" => "application/json"}
                       (if (:auth client) {"Authorization" => (:auth client)} {}))
        response (client/post (:url client)
                              {:headers headers
                               :body payload
                               :throw-exceptions false})
        status (:status response)
        body (json/parse-string (:body response) true)]
    (if (and (not= status 200) (not= status 500))
      (throw (Exception. (str "HTTP Error: Status " status)))
      (if-let [err (:error body)]
        (throw (ex-info (str "RPC Error [" (:code err) "]: " (:message err))
                        {:type :raptoreum/rpc-error
                         :code (:code err)
                         :message (:message err)}))
        (:result body)))))

(defn get-blockchain-info [client]
  (request client "getblockchaininfo"))

(defn get-block-count [client]
  (request client "getblockcount"))

(defn get-balance [client]
  (request client "getbalance"))

(defn validate-address [client address]
  (request client "validateaddress" [address]))

(defn send-many [client amounts & [minconf comment]]
  (request client "sendmany" ["" amounts (or minconf 1) (or comment "")]))

(defn list-assets [client & [mine]]
  (request client "listassets" [(or mine false)]))

(defn create-asset [client name amount & [options]]
  (request client "createasset" [name amount (or options {})]))

(defn send-asset [client asset-id amount address]
  (request client "sendasset" [asset-id amount address]))

File: sdk/julia/src/Raptoreum.jl
module Raptoreum

using HTTP
using JSON3
using Base64

export RaptoreumClient, request, getblockchaininfo, getblockcount, getbalance

struct RaptoreumClient
    url::String
    headers::Vector{Pair{String, String}}
end

function RaptoreumClient(host::String="127.0.0.1", port::Int=8766, user::String="", password::String="", use_ssl::Bool=false)
    scheme = use_ssl ? "https" : "http"
    url = "$scheme://$host:$port/"
    
    headers = ["Content-Type" => "application/json"]
    if !isempty(user) || !isempty(password)
        auth = base64encode("$user:$password")
        push!(headers, "Authorization" => "Basic $auth")
    end
    
    return RaptoreumClient(url, headers)
end

function request(client::RaptoreumClient, method::String, params::Vector=Any[])
    payload = Dict(
        "jsonrpc" => "1.0",
        "id" => "rtm-sdk-julia",
        "method" => method,
        "params" => params
    )
    
    body = JSON3.write(payload)
    
    try
        response = HTTP.post(client.url, client.headers, body, status_exception=false)
        if response.status != 200 && response.status != 500
            error("HTTP Error: Received status code $(response.status)")
        end
        
        parsed = JSON3.read(String(response.body))
        if haskey(parsed, :error) && parsed.error !== nothing
            error("RPC Error [$(parsed.error.code)]: $(parsed.error.message)")
        end
        
        return parsed.result
    catch e
        rethrow(e)
    end
end

getblockchaininfo(client::RaptoreumClient) = request(client, "getblockchaininfo")
getblockcount(client::RaptoreumClient) = request(client, "getblockcount")
getbalance(client::RaptoreumClient) = request(client, "getbalance")

end # module
